Date/time formatting simplifications

labkey-adam opened the pull request:
#### Rationale
Server-side UI changes for https://docs.google.com/document/d/1XcatQe9wRDw1ONLML9x4fFv_fIq9zhXPHoYANBuOykw/edit#heading=h.4rg4h5vuek0

#### Related Pull Requests
- https://github.com/LabKey/platform/pull/5841

- https://github.com/LabKey/labkey-api-js/pull/184
- https://github.com/LabKey/platform/pull/5923
- https://github.com/LabKey/labkey-ui-components/pull/1607
- https://github.com/LabKey/labkey-ui-premium/pull/563
- https://github.com/LabKey/limsModules/pull/826
- https://github.com/LabKey/testAutomation/pull/2099
- https://github.com/LabKey/premiumModules/pull/76
- https://github.com/LabKey/ehrModules/pull/844
- https://github.com/LabKey/snprcEHRModules/pull/762

#### Changes
- Select date-time formats via drop-down lists holding standard formats (and non-standard, if already set)
- Add "Inherit" checkboxes to allow inheritance/override of individual properties (allowed previously for a few properties via hidden UI)
- Fix broken `LookAndFeelProperties` methods that ostensibly retrieved the stored value but actually retrieved inherited values. (Calling the parent class lookup method with a enum inadvertently walked the hierarchy.) Export and inheritance UI should be correct for all settings now.
- Simplify `SettingsForm` / `FolderSettingsForm` / `ProjectSettings` hierarchy: remove pointless SettingsForm interface and have ProjectSettings extend FolderSettings to reduce boilerplate
- Update `saveProjectSettings()` to take a form instead of JSON

labkey-adam committed:
Date/time formatting simplifications

labkey-adam committed:
Add inheritance UI

labkey-adam committed:
Persist inheritance based on new posted values, handle NONE properly, allow empty parsing patterns, fix splitting

labkey-adam committed:
Comments

labkey-adam committed:
Merge remote-tracking branch 'origin/develop' into fb_datetime_formats

labkey-adam committed:
Fix restrict charting columns inheritance

labkey-adam committed:
Warnings

labkey-adam committed:
Display inputs for inherited properties. Allow empty number format.

labkey-adam committed:
Inherited help pop-up

labkey-adam committed:
Eliminate SettingsForm interface and remove duplicate getters & setters from ProjectSettingsForm

labkey-adam committed:
Migrate UpdateProjectSettingsAction to use FolderSettingsForm and saveFolderSettings()

labkey-adam committed:
Merge remote-tracking branch 'origin/develop' into fb_datetime_formats

labkey-adam committed:
Checkpoint for project settings migration

labkey-adam committed:
System description, short name, and theme

labkey-adam committed:
Folder navigation, app menu, help menu, discussions

XingY committed:
App UI

XingY committed:
fix site settings

labkey-adam committed:
Finish project settings

labkey-adam committed:
Clean up. Show all errors instead of one at a time.

labkey-adam committed:
Unused import

labkey-adam committed:
Merge remote-tracking branch 'origin/develop' into fb_datetime_formats

XingY committed:
Merge branch 'develop' into fb_datetime_formats

XingY committed:
Merge branch 'develop' into fb_datetime_formats

# Conflicts:
#	biologics/package-lock.json
#	biologics/package.json
#	inventory/package-lock.json
#	inventory/package.json
#	sampleManagement/package-lock.json
#	sampleManagement/package.json

XingY committed:
update style

labkey-adam committed:
Simplify inheritCheckbox() and clarify "Reset" button caption & confirmation message

cnathe commented on api/src/org/labkey/api/data/Container.java:
```diff
@@ -1384,10 +1385,27 @@ public Map<String, Object> toJSON(User user, boolean includePermissions, boolean
             LookAndFeelProperties props = LookAndFeelProperties.getInstance(this);
             Map<String, Object> formats = new HashMap<>();
-            formats.put("dateFormat", DateUtil.getDateFormatString(this));
+            formats.put("dateFormat", props.getDefaultDateFormat()); // when to use DateUtil.getDateFormatString vs laf.getDefaultDateFormat?
```
is this comment a TODO for this PR or for a future refactor / change?

XingY replied:
I added this comment and hope someone with more knowledge could provide an answer. Historically the usage have been inconsistent.

labkey-adam replied:
Interesting... I would have thought that `props.getDefaultDateFormat()` would be the preferred method, but it always walks the container hierarchy whereas the `DateUtil` method uses the cache. I'll take a closer look at this... we may want to adjust LookAndFeelProperties to reference the cache since it's used extensively.

labkey-adam replied:
These methods returned identical results, but one used the cache and one didn't. I did a mini-refactor on `LookAndFeelFolderProperties` so it delegates to the cache as well, except for new methods (e.g., `calculateDefaultDateFormat()`) called only by the cache loader. `DateUtil.getDateFormatString()` and `props.getDefaultDateFormat()` are now identical; callers can choose one or the other based solely on convenience (code needing just a single format would likely call `DateUtil`; code needing multiple properties would likely use LAFFP).

cnathe commented on api/src/org/labkey/api/data/Container.java:
```diff
@@ -1384,10 +1385,27 @@ public Map<String, Object> toJSON(User user, boolean includePermissions, boolean
+
+                if (!isRoot())
+                {
+                    Container parentContainer = getParent();  // when to use DateUtil.getDateFormatString vs laf.getDefaultDateFormat?
```
same question, is this comment meant for a TODO in this PR?

labkey-adam replied:
I'll take a look

cnathe commented on api/src/org/labkey/api/settings/LookAndFeelProperties.java:
```diff
@@ -154,33 +158,60 @@ public static WriteableFolderLookAndFeelProperties getWriteableFolderInstance(Co
+
+    public DateParsingMode getDateParsingMode()
+    {
+        assert _settingsContainer.isRoot();
```
looks like this assert is only in this one getter method. should it be in `getCustomWelcome` as well (and maybe others)?

labkey-susanh replied:
Should this be an explicit check that throws an exception (or the like) so we don't do the wrong thing if assertions are not enabled?

XingY replied:
I added this assert originally. Agree it should be updated to exception.

cnathe commented on api/src/org/labkey/api/settings/LookAndFeelProperties.java:
```diff
@@ -192,46 +223,113 @@ public String getThemeName()
+        // Prefer correctly spelled property name, but fall-back to the old, misspelled one
+        String enabled = getStoredValue(discussionEnabled);
+        if (enabled == null)
+            enabled = getStoredValue(_settingsContainer, "dicussionEnabled");
```
bummer to have to keep the fallback to the misspelled property name. is that something we could just fix in an upgrade script and then stop supporting here?

labkey-adam replied:
I would do that, but discussions (and this setting) are deprecated and likely going away soon, so not worth the bother.

cnathe commented on core/src/org/labkey/core/admin/AdminController.java:
```diff
@@ -5504,25 +5556,69 @@ public void validateCommand(FolderSettingsForm form, Errors errors)
+        validateAndSaveFormat(form.getDefaultDateTimeFormat(), form.isDefaultDateTimeFormatInherited(), props::clearDefaultDateTimeFormat, props::setDefaultDateTimeFormat, errors, "date-time");
+        validateAndSaveFormat(form.getDefaultTimeFormat(), form.isDefaultTimeFormatInherited(), props::clearDefaultTimeFormat, props::setDefaultTimeFormat, errors, "time");
+        validateAndSaveFormat(form.getDefaultNumberFormat(), form.isDefaultNumberFormatInherited(), props::clearDefaultNumberFormat, props::setDefaultNumberFormat, errors, "number");
+        validateAndSaveFormat(form.getExtraDateParsingPattern(), form.isExtraDateParsingPatternInherited(), props::clearExtraDateParsingPattern, props::setExtraDateParsingPattern, errors, "date");
+        validateAndSaveFormat(form.getExtraDateTimeParsingPattern(), form.isExtraDateTimeParsingPatternInherited(), props::clearExtraDateTimeParsingPattern, props::setExtraDateTimeParsingPattern, errors, "date-time");
+        validateAndSaveFormat(form.getExtraTimeParsingPattern(), form.isExtraTimeParsingPatternInherited(), props::clearExtraTimeParsingPattern, props::setExtraTimeParsingPattern, errors, "time");
```
do we want the "what" last parameter to be distinquishable for these "extra parsing" options? right now these match the lines 5566 through 5569 above (but maybe we don't care since this is just used in the error message).

labkey-adam replied:
`what` is now more specific in all cases

cnathe commented on core/src/org/labkey/core/admin/lookAndFeelProperties.jsp:
```diff
@@ -43,36 +54,50 @@
+    }
+    HtmlString clearMessage = HtmlString.of(
+        c.isRoot() ? "all look & feel properties to their default values" : (
+        c.isProject() ? "all project settings to inherit from the site look & feel properties" :
```
the `&` here ends up showing up as `&amp;` in the FF and Chrome modal:
<img width="531" alt="Screenshot 2024-10-15 at 9 54 38 AM" src="https://github.com/user-attachments/assets/3923ed00-7528-4cc6-a9e6-1e4f0eff3302">

labkey-adam replied:
Double encoding fixed

cnathe commented on core/src/org/labkey/core/admin/lookAndFeelProperties.jsp:
```diff
@@ -95,107 +120,133 @@
+    %>
+        <td style="padding-left: 5px; padding-right: 5px;">Inherited<%=inheritHelp%></td>
+    <%
+        }
```
should this have an else that writes out an empty `<td>` so that the colspan of 3 matches? or should the colspan be conditional on `c.isProject()`?

labkey-adam replied:
This <td> didn't have a colspan, so I added it conditionally (site case)

cnathe commented on core/src/org/labkey/core/admin/lookAndFeelProperties.jsp:
```diff
@@ -43,36 +54,50 @@
+    String currentThemeName = laf.getThemeName();
     boolean canUpdate = !c.isRoot() || c.hasPermission(getUser(), ApplicationAdminPermission.class);
     boolean hasPremiumModule = ModuleLoader.getInstance().hasModule("Premium");
+    int standardInputWidth = 60;
 %>
 <%=formatMissedErrors("form")%>
+<div id="dateFormatWarning" style="display: none;" class="alert alert-warning alert-dismissable">
+    <div class="lk-dismissable-warn">Warning: One or more date, time, or date-time display formats are using non-standard patterns. <%=helpLink("studyDateNumber", "Click here")%> to learn more.</div>
+</div>
 <labkey:form name="preferences" method="post" id="form-preferences">
 <table class="lk-fields-table">
-<%=getTroubleshooterWarning(canUpdate, HtmlString.unsafe("<tr><td colspan=2>"), HtmlString.unsafe("</td></tr>"))%>
+<%=getTroubleshooterWarning(canUpdate, HtmlString.unsafe("<tr><td colspan=3>"), HtmlString.unsafe("</td></tr>"))%>
 <tr>
-    <td colspan=2>&nbsp;</td>
+    <td colspan=3>&nbsp;</td>
 </tr>
 <%
     if (c.isProject())
     {
 %>
 <tr>
-    <td colspan=2>Security defaults</td>
+    <td colspan="3">Security defaults</td>
 </tr>
 <tr>
     <td class="labkey-form-label"><label for="shouldInherit">New folders should inherit permissions by default</label></td>
-    <td><input type="checkbox" id="shouldInherit" name="shouldInherit" size="50"<%=checked(SecurityManager.shouldNewSubfoldersInheritPermissions(c))%>></td>
+    <td><input type="checkbox" id="shouldInherit" name="shouldInherit" size="<%=standardInputWidth%>"<%=checked(SecurityManager.shouldNewSubfoldersInheritPermissions(c))%>></td>
```
for consistency in the UI, should we add an empty `<td>` here so that the checkbox lines up with the 3rd table column (since this property doesn't have the inherited checkox).

labkey-adam replied:
Sure

labkey-adam replied:
Added

cnathe commented on core/src/org/labkey/core/admin/lookAndFeelProperties.jsp:
```diff
@@ -408,4 +499,109 @@
+        }
+
+        // Add a non-standard and an invalid format to every drop-down for testing purposes
+        // TODO: Delete this block - just for testing
+        Set<String> set = new LinkedHashSet<>(options);
+        set.add(type == DateDisplayFormatType.Date ? "MM-dd-yyyy" : "kk:mm"); // non-standard
+        set.add("xyz"); // invalid
```
just adding a note here to make sure this is removed after testing / before merge.

labkey-adam replied:
Removed

XingY committed:
Merge branch 'develop' into fb_datetime_formats

XingY committed:
code review changes

XingY committed:
code review changes

XingY committed:
jest

labkey-adam commented on api/src/org/labkey/api/data/Container.java:
```diff
@@ -1384,10 +1385,27 @@ public Map<String, Object> toJSON(User user, boolean includePermissions, boolean
+
+            if (includeParentFormatProps)
+            {
+                formats.put("dateFormatInherited", StringUtils.isEmpty(props.getDefaultDateFormatStored()));
```
I'll update these, but I think they should be checking for null, not empty. Number format is often set to "", which doesn't mean inherit.

labkey-adam committed:
Code review feedback. LookAndFeelFolderProperties delegate to the folder settings cache.

labkey-adam committed:
Double encoding is better than no encoding... but single encoding is even better

labkey-adam committed:
Extra <td>. Extra colspan.

labkey-adam committed:
Update splitting and handling of non-standard date-time formats

labkey-adam committed:
Display full non-standard date-time in date select

labkey-danield committed:
Merging develop into fb_datetime_formats.

XingY committed:
Merge branch 'develop' into fb_datetime_formats

XingY committed:
Merge branch 'develop' into fb_datetime_formats

# Conflicts:
#	biologics/package-lock.json
#	biologics/package.json
#	inventory/package-lock.json
#	inventory/package.json
#	sampleManagement/package-lock.json
#	sampleManagement/package.json

XingY committed:
Merge branch 'develop' into fb_datetime_formats

# Conflicts:
#	biologics/package-lock.json
#	biologics/package.json
#	inventory/package-lock.json
#	inventory/package.json
#	sampleManagement/package-lock.json
#	sampleManagement/package.json

labkey-adam committed:
Junit tests for DateUtil methods operating on standard and non-standard formats

labkey-adam committed:
Junit test for inheritance

labkey-danield committed:
Fixing existing test using date and/or time formats.
Changed settings page to take an enum for the date and time formats.
Added date and time formats as enums to BaseSettingsPage.

labkey-adam committed:
Remove test formats

labkey-adam committed:
Merge remote-tracking branch 'origin/develop' into fb_datetime_formats

labkey-danield committed:
Merge branch 'develop' into fb_datetime_formats

XingY committed:
Merge branch 'develop' into fb_datetime_formats

XingY committed:
Merge branch 'develop' into fb_datetime_formats

labkey-danield committed:
Fix existing test to work with new DateTime and "inherited" functionality.

labkey-danield committed:
Merge remote-tracking branch 'origin/fb_datetime_formats' into fb_datetime_formats

labkey-adam committed:
Fix RowIds with commas

labkey-danield committed:
Check the right thing.

labkey-danield committed:
Merge remote-tracking branch 'origin/fb_datetime_formats' into fb_datetime_formats

XingY committed:
Merge branch 'develop' into fb_datetime_formats

# Conflicts:
#	core/package-lock.json
#	core/package.json
#	experiment/package-lock.json
#	experiment/package.json

XingY committed:
update ver
